numpyro/infer/reparam.py

```diff
@@ -26,6 +26,93 @@ def __call__(self, name, fn, obs):
         """
         return fn, obs
 
+    def _unwrap(self, fn):
+        """
+        Unwrap Independent(...) distributions.
+        """
+        event_dim = fn.event_dim
+        while isinstance(fn, dist.Independent):
+            fn = fn.base_dist
+        return fn, event_dim
+
+    def _wrap(self, fn, event_dim):
+        """
+        Wrap in Independent distributions.
+        """
+        if fn.event_dim < event_dim:
+            fn = fn.to_event(event_dim - fn.event_dim)
+        assert fn.event_dim == event_dim
+        return fn
+
+    def _unexpand(self, fn):
+        """
+        Unexpand ExpandedDistribution(...) distributions.
+        """
+        batch_shape = fn.batch_shape
+        if isinstance(fn, dist.ExpandedDistribution):
+            fn = fn.base_dist
+        return fn, batch_shape
+
+
+class LocScaleReparam(Reparam):
+    """
+    Generic decentering reparameterizer [1] for latent variables parameterized
+    by ``loc`` and ``scale`` (and possibly additional ``shape_params``).
+
+    This reparameterization works only for latent variables, not likelihoods.
+
+    **References:**
+
+    1. *Automatic Reparameterisation of Probabilistic Programs*,
+       Maria I. Gorinova, Dave Moore, Matthew D. Hoffman (2019)
+
+    :param float centered: optional centered parameter. If None (default) learn
+        a per-site per-element centering parameter in ``[0,1]``. If 0, fully
+        decenter the distribution; if 1, preserve the centered distribution
+        unchanged.
+    :param shape_params: list of additional parameter names to copy unchanged from
+        the centered to decentered distribution.
+    :type shape_params: tuple or list
+    """
+    def __init__(self, centered=None, shape_params=()):
+        assert centered is None or isinstance(centered, (int, float))
```

Comment:
@fritzo do you use Tensor `centered`? In NumPyro, under `jit` context, we can't check if `centered` belongs to the interval [0, 1] so I skip the support for it. If you think it is useful, I'll remove this assertation.

Reply:
Yes, the main idea of the [original paper](https://arxiv.org/abs/1906.03028) is to learn per-element `centered` parameters. If you simplify anything, I would simplify to make `centered` required and to always be a tensor and assume it is learned.

Reply:
Oh I see, you call `param` statements internally. I think this is a fine assumption.